caffe2/core/registry.h

```diff
@@ -158,7 +158,8 @@ class Registerer {
 
 #define CAFFE_DEFINE_TYPED_REGISTRY(                                         \
     RegistryName, SrcType, ObjectType, PtrType, ...)                         \
-  Registry<SrcType, PtrType<ObjectType>, ##__VA_ARGS__>* RegistryName() {    \
+  CAFFE2_EXPORT Registry<SrcType, PtrType<ObjectType>, ##__VA_ARGS__>*       \
```

Comment:
So with the `CAFFE2_EXPORT` I'm assuming this only gets called in `.cc` or `.cpp` files? Just trying to understand the choice.

Reply:
Ah yeah, so basically, the choice of API (which at build time adapts to IMPORT or EXPORT), EXPORT and IMPORT are as follows:

(1) for function and class definitions in header files: always use API. This will then be routed at compile time: when we are building the main library, because we set CAFFE2_BUILD_MAIN_LIB, it routes us to EXPORT, allowing the definition in the .cc files to be exported. When we are using the headers in a dependent library (someone that uses Caffe2/pytorch), then because CAFFE2_BUILD_MAIN_LIB is not set, it routes to IMPORT, which tells the compiler the function definitions are going to be elsewhere.

(2) in .cc files where we actually define the variables (global variables, static variables and functions), explicitly using EXPORT is safer because it avoid programming error.

(3) in header files when defining macros, and if the macro is going to be used in cc files for global variable definition (such as this line) - explicitly use EXPORT. The reason is that, such lines may be used in dependent libraries too (imagine a dependent library defining a new registry for a new device) - and when they do so, they want to explicitly EXPORT the registry.

Reply:
Put it in a comment?

Reply:
@ezyang Thanks, will do - maybe after we merge all the macros... Right now it is kinda everywhere, and AT_API doesn't have export/import equivalent yet.